examples/src/main/java/io/dapr/examples/actors/DemoActorClient.java

```diff
@@ -30,7 +31,8 @@
    * @throws InterruptedException If program has been interrupted.
    */
   public static void main(String[] args) throws InterruptedException {
-    try (ActorProxyBuilder<DemoActor> builder = new ActorProxyBuilder(DemoActor.class)) {
+    try (DaprChannel channel = new DaprChannel()) {

```

Comment:
This is why this is a breaking change. This is only applicable for Actors because the proxy is per actor instance, so there must be a "client" shared some how. I think having this separate class is more intuitive then making the ActorProxyBuilder to own the channel behind the scenes.

Reply:
Two questions about this approach...

---

I assume that it's not feasible to make the generated proxies (`actor` in this example) implement `Closable` because it's an interface that the user provided and may not include it.  

So having the user do the equivalent of (my Java is rusty, be nice!):

```java
if (actor instanceof Closable) {
    ((Closable)actor).Close()
}
```

Just sounds terrible from a usability standpoint. A beginner would not understand what to do.

Equally bad is requiring the actor interface to implement `Closable` because then they end up with a *do-nothing* implementation of that interface on the server, since the proxy interface is shared.

---

Second question.

Have you considered using a static instance of `DaprChannel` and making the constructor arg optional via overloading? This way by default users can share a single channel instance that's process-scoped and can opt-in to more complexity if they want eager resource cleanup.

---

I also considered the approach of returning a *handle* type, which would not only be a massive breaking change, but it would also require ref-counting (yuck).

Reply:
First question, making the proxy or the class user provided to be closeable does not make sense because the resource to be closed (grpcChannel) is shared and reused. So a single instance of ActorProxy cannot close it because there might be other ActorProxies still using it.

The idea is to give users freedom of where to create and when to close the managed resource. If I create one for them. Having a static instance will also load it even when users don't use it. Imagine having this managed channel created for the application that just serves the actor runtime without calling into other actors? Users are also used to do this in Springboot - actually being opinionated about resource management probably belongs to the springboot package we have instead of here.

No ref counting :)

Reply:
> Users are also used to do this in Springboot - actually being opinionated about resource management probably belongs to the springboot package we have instead of here.

So does this mean we have a recipe for using these types in a DI-enabled way? Or is that up the user to figure out?